Python 3 Compatibility Fixes

jdeschenes opened the pull request:
As per my issue in the [chaco issue]https://github.com/enthought/chaco/issues/368)

This is a PR to completely get rid of the 2to3 using the six library

Most of the fixes are pretty mechanical and should be pretty straightforward to review:

* print statement to print function
* new raise syntax
* relative imports
* whenever I could, I converted the list iteration to an iterator
   * items(), keys(), values() to six.iteritems(), iterkeys() and itervalues() (Some of the loops are playing with 
the list, dictionary in the loop and those have been explicitely converted)
    * range to xrange

* LONG_TYPE instead of long
* six.u instead of u''
* and a lot more

The branch has been tested on python 2.7 and python 3.6.

jwiggins commented:
Is `six.u` really necessary if the supported Python versions are 2.7 and 3.4+?

mdickinson commented on examples/tutorials/doc_examples/examples/all_wildcard.py:
```diff
@@ -32,11 +32,11 @@ class Person(HasTraits):
-except TraitError, c:
-    print 'TraitError: ', c, '\n'
+except TraitError as c:
+    print('TraitError: ', c, '\n')
```
This one should probably be changed to use formatting, otherwise it's doing something different on Python 2 (printing a `tuple` rather than individual items separated by spaces). Alternatively, we could add a `from __future__ import print_function` to the top of the module.

mdickinson replied:
Resolved via the `print_function` import; thanks!

mdickinson commented on examples/tutorials/doc_examples/examples/delegate.py:
```diff
@@ -26,28 +26,28 @@ class Child(HasTraits):
-except TraitError, c:
-    print 'TraitError: ', c
+except TraitError as c:
+    print('TraitError: ', c)
```
Another case where either formatting or a `__future__` import should be used. I'll stop commenting on these at this point ...

mdickinson commented on examples/tutorials/tutor.py:
```diff
@@ -766,7 +768,7 @@ def run_code(self):
 
                 # Execute the current lab code:
-                exec module[2:] in values, values
+                six.exec_(module[2:] in values, values)
```
I think we probably don't need `six` here: `exec(module[2:], values, values)` should work on both Python 2 and Python 3.

jdeschenes replied:
Done

mdickinson commented on traits/adaptation/adaptation_manager.py:
```diff
@@ -319,7 +321,7 @@ def _get_applicable_offers(self, current_protocol, path):
         edges = []
 
-        for from_protocol_name, offers in self._adaptation_offers.items():
+        for from_protocol_name, offers in six.iteritems(self._adaptation_offers):
```
I'd suggest just leaving this one as it is. I think it should be fine on both Python 2 and Python 3. (We end up computing an actual list in Python 2, and just iterating over a view in Python 3, but I don't think that matters.)

mdickinson replied:
I'll merge as is; we can fix later.

mdickinson commented on traits/adaptation/tests/benchmark.py:
```diff
@@ -27,15 +31,15 @@ class IFoo{i}(Interface):
 class Foo{i}(HasTraits):
     pass
 """
-for i in range(N_SOURCES):
```
I think it's fine to leave all these as `range`. It would be more efficient to be using `xrange` in Python 2, but it would also be a change from the previous behaviour, and we don't need to do it to make this code work under both Python 2 and Python 3.

mdickinson commented on traits/category.py:
```diff
@@ -88,6 +92,8 @@ def add_traits_meta_data ( self, bases, class_dict, base_traits,
 #-------------------------------------------------------------------------------
 
+
+@six.add_metaclass(MetaCategory)
```
Would this look neater using `with_metaclass`?  `class Category(with_metaclass(MetaCategory, object))`?

jdeschenes replied:
Personally, I prefer the decorator variant as I find it easier to read. 

That being said, here is a gotcha when using the decorator:

```python
import six

class Meta1(type):
    pass

class Meta2(type):
    pass

class CombinedMeta(Meta1, Meta2):
    pass

@six.add_metaclass(Meta1)
class Base1(object):
    pass

@six.add_metaclass(Meta2)
class Base2(object):
    pass

# This works as intended
class DerivedClass(six.with_metaclass(CombinedMeta, Base1, Base2)):
    pass

# This fails with a TypeError(metaclass conflict)
@six.add_metaclass(CombinedMeta)
class DerivedClass(Base1, Base2):
    pass
```
I found this while running your test suite(test_abc.py). As far as I am aware, there is no impact once the class is created.

In the six documentation there is nothing that mentions anything like that.

mdickinson replied:
Right, I guess that's because the class decorator only has a chance to do its thing _after_ the class has been created. I haven't looked into the `six.with_metaclass` source to see what magic it uses to avoid this problem.

mdickinson commented on traits/etsconfig/etsconfig.py:
```diff
@@ -266,8 +266,8 @@ def _set_toolkit(self, toolkit):
         if self._toolkit and self._toolkit != toolkit:
-            raise ValueError, "cannot set toolkit to %s because it has "\
-                            "already been set to %s" % (toolkit, self._toolkit)
+            raise ValueError("cannot set toolkit to %s because it has "\
```
The trailing backslash here could be removed.

mdickinson commented on traits/has_dynamic_views.py:
```diff
@@ -199,7 +201,7 @@ def trait_view ( self, name = None, view_element = None ):
             # dynamic views should be the default view:
             if (view_element is None) and (name is None or len( name ) < 1):
-                for dname, declaration in self._dynamic_view_registry.items():
+                for dname, declaration in six.iteritems(self._dynamic_view_registry):
```
I think it should be fine to leave this one unchanged.

mdickinson commented on traits/has_dynamic_views.py:
```diff
@@ -336,8 +338,7 @@ def _get_dynamic_elements ( self, name ):
         # Sort the contributed elements by their display ordering values:
-        ordering = filtered.keys()
-        ordering.sort()
+        ordering = sorted(six.iterkeys(filtered))
```
Just `ordering = sorted(filtered)` should be enough here.

mdickinson commented on traits/has_traits.py:
```diff
@@ -242,7 +223,7 @@ def _is_serializable ( value ):
 
     if isinstance( value, dict ):
-        for name, item in value.items():
+        for name, item in six.iteritems(value):
```
Again, I think it's fine to leave this as-is.

mdickinson commented:
Many thanks for this! This is very valuable, and I hope we can integrate it soon. I've taken a look through the first half of the diff, and left some comments; I'll try to find time to complete the review soon.

Some general comments so far:
- I think there are a lot of cases where we don't need to use `six`: in particular, code like `for k, v in some_dict.items()` can probably be left unchanged.
- Where a `print` statment is changed from `print a, b` to `print(a, b)`, that changes the meaning on Python 2: the new code prints a tuple (with parentheses and commas, and repr-formatted items) where the old code printed space-separated str-formatted items. Those cases probably need to be replaced with string formatting. Example:
```
>>> import datetime
>>> time_now = datetime.datetime.utcnow()
>>> print "Time now is:", time_now   # old behaviour on Python 2
Time now is: 2017-07-21 08:08:19.598678
>>> print("Time now is:", time_now)  # new behaviour on Python 2
('Time now is:', datetime.datetime(2017, 7, 21, 8, 8, 19, 598678))
>>> print("Time now is: {}".format(time_now))  # proposed replacement
Time now is: 2017-07-21 08:08:19.598678
```
- As @jwiggins says: we probably don't need `six.u`, since `u`-prefixes work in Python 3.3+, and we're not planning to support Python 3.2. (And probably not Python 3.3, either.)

mdickinson commented:
> Those cases probably need to be replaced with string formatting

Using `from __future__ import print_function` would probably be simpler.

jdeschenes commented:
Hi,
Some changes:
* I had completelely forgotten about the print function difference in python 2.7. It's now fixed.
* I removed all six.u references
* I made some other very small changes

For the items() vs iteritems(), I went over all the cases and changed it manually wherever possible to the iteritems() variant as it is more efficient. It also has the bigger benefit of having the same behavior in both python 2 and 3. I would like to keep it in there, but I can understand that this is a functional change. Would you be open to have this maybe in a separate PR?

mdickinson commented on docs/source/_extensions/refactordoc/line_functions.py:
```diff
@@ -73,7 +74,7 @@ def trim_indent(lines):
 
     """
-    non_empty_lines = filter(lambda x: not is_empty(x), lines)
+    non_empty_lines = sm.filter(lambda x: not is_empty(x), lines)
```
I think it's better to leave this as regular `filter`. To do a bit of a brain dump: there are lots of tradeoffs here: on the one hand, consistent behaviour between Python 2 and Python 3 is a good thing, and it's nice to get rid of a minor inefficiency in Python 2. On the other hand, any change to the Python 2 code carries a small risk of code breakage (we have _huge_ amounts of Python 2 code that depends on Traits, and much less Python 3 code), the inefficient form was already working, this isn't a critical code path anyway, and the `sm.filter` looks ugly.

My own weighted average of all those considerations leads me to a "leave it as is".

mdickinson commented on docs/source/_extensions/refactordoc/line_functions.py:
```diff
@@ -8,7 +8,8 @@
 #-----------------------------------------------------------------------------
 import re
-
+import six
```
It looks as though this import is unused..

mdickinson commented on examples/tutorials/traits_4.0/extended_trait_change/properties.py:
```diff
@@ -124,7 +127,7 @@ class LeagueModelView(ModelView):
         """ Returns the total number of hits across all teams and players.
         """
-        return reduce(add, [reduce(add, [p.hits for p in t.players], 0)
+        return sm.reduce(add, [sm.reduce(add, [p.hits for p in t.players], 0)
```
Gah. I'm not sure why we're so fond of `reduce`-based examples. I have a suspicion that much of this dates from before Python 2.5, when `sum` was introduced. Still, that's not something to change in this PR.

jdeschenes commented:
I went over the code. From the commit log:

* I removed all unused `import six`
* I fixed what you specifically asked

If you want more changes for the iteritems... Can you mention the approach you want to have?

jdeschenes commented:
@mdickinson Let me know if there  is anything else you would like to be done!

jdeschenes commented:
@mdickinson All iteritems, itervalues and iterkeys were removed wherever possible. Let me know if there is anything else to do.

jdeschenes commented:
I just submitted a fix. I think that this test may be skipped in python 3 altogether instead of what I did. Let me know what you prefer.

mdickinson commented:
Thanks for all the updates; I've been short on cycles to look at this recently, but hope to fix that in the next couple of days.

jdeschenes commented:
No problem! Quick question... Why did that appveyor failure only pop up on radar today?

mdickinson commented on traits/trait_base.py:
```diff
@@ -178,18 +180,18 @@ def strx ( arg ):
 # Mapping of coercable types.
 CoercableTypes = {
-    long:    ( 11, long, int ),
+    six.integer_types:    ( 11, six.integer_types, int ),
```
This doesn't look right: `six.integer_types` is a tuple, and I think we want an expanded list of types here.

mdickinson commented on traits/trait_types.py:
```diff
@@ -167,7 +175,7 @@ def validate ( self, object, name, value ):
         if type(value) is int:
             return value
-        elif type(value) is long:
+        elif type(value) is six.integer_types:
```
I think this line should be `elif type(value) in six.integer_types:`

mdickinson commented on traits/trait_base.py:
```diff
@@ -51,7 +53,7 @@
 ComplexTypes  = ( float, int )
 
-TypeTypes     = ( str,  unicode, int, long, float, complex, list, tuple, dict, bool )
+TypeTypes     = ( str,  six.text_type, int, six.integer_types, float, complex, list, tuple, dict, bool )
```
`six.integer_types` needs to be expanded here

mdickinson commented:
Code changes look good, overall. I think some of the uses of `six.integer_types` need to be changed, but apart from that I couldn't spot any issues.

I plan to run one of our large, well-tested, Traits-using applications against this branch, as a smoke test. Once that's done, and the `six.integer_types` uses fixed, I think this is ready to go.

mdickinson commented:
> Why did that appveyor failure only pop up on radar today?

Our Appveyor configuration was accidentally disabled on this repository for a while; I notice this last week and fixed it.

jdeschenes commented:
I have fixed dubious uses of six.integer_types.

mdickinson commented:
> I plan to run one of our large, well-tested, Traits-using applications against this branch, as a smoke test.

Done, with a number of test failures. I'm rerunning, since I wasn't running against the latest commit, and will investigate and report back.

mdickinson commented:
Here's the bug that appears to account for most of the test failures:
```
(geophysics)bash-3.2$ python
Enthought Deployment Manager -- https://www.enthought.com
Python 2.7.13 |Enthought, Inc. (x86_64)| (default, Mar  2 2017, 08:20:50) 
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from traits.api import *
>>> class A(HasTraits):
...     my_dir = Directory
... 
>>> a = A()
>>> import os                         
>>> a.my_dir = os.path.expanduser("~")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/mdickinson/Desktop/traits/traits/trait_handlers.py", line 177, in error
    value )
traits.trait_errors.TraitError: The 'my_dir' trait of an A instance must be a directory name, but a value of '/Users/mdickinson' <type 'str'> was specified.
```

mdickinson commented on traits/has_traits.py:
```diff
@@ -1037,7 +1020,7 @@ class HasTraits ( CHasTraits ):
 
     #: An event fired when a new trait is dynamically added to the object
-    trait_added = Event( basestring )
+    trait_added = Event( six.string_types )
```
I think this needs to be `basestring` on Python 2 and `str` on Python 3. I'm not sure that using a `tuple` of types in an `Event` works.

mdickinson commented on traits/trait_types.py:
```diff
@@ -1532,7 +1533,7 @@ def __init__ ( self, value = '', filter = None, auto_set = False,
         if not exists:
             # Define the C-level fast validator to use:
-            fast_validate = ( 11, basestring )
+            fast_validate = ( 11, six.string_types )
```
Need `six.string_types[0]` here.

mdickinson commented on traits/trait_types.py:
```diff
@@ -1622,7 +1623,7 @@ def __init__ ( self, value = '', auto_set = False, entries = 0,
         #: test is not required:
         if not exists:
-            self.fast_validate = ( 11, basestring )
+            self.fast_validate = ( 11, six.string_types )
```
`six.string_types[0]`

mdickinson commented on traits/trait_types.py:
```diff
@@ -1673,14 +1674,14 @@ def __init__ ( self, low = None, high = None, value = None,
         vtype = type( high )
         if ((low is not None) and
-            (not issubclass( vtype, ( float, basestring ) ))):
+            (not issubclass( vtype, ( float, six.string_types ) ))):
```
The nested tuple needs denesting.

mdickinson replied:
Fixed locally; will push shortly.

mdickinson commented:
It looks as though the uses of `six.string_types` need re-examining. I've left individual comments on all the instances I found that need changing, but I may have missed some.

It's a little odd that `six.string_types` is _always_ a tuple of length 1; I wonder why Benjamin didn't go with a simple type instead. Perhaps for consistency with `six.integer_types`?

mdickinson commented on traits/trait_types.py:
```diff
@@ -1673,14 +1674,14 @@ def __init__ ( self, low = None, high = None, value = None,
             vtype = type( low )
 
-        is_static = (not issubclass( vtype, basestring ))
+        is_static = (not issubclass( vtype, six.string_types ))
```
Denesting needed here, too.

mdickinson replied:
Nope, sorry; my mistake. This one's fine.

jdeschenes commented:
I fixed the issues you mentionned. I have taken the liberty to add a few test cases to test what was not covered. In doing so, I found a bug with the Directory trait. I fixed it as well.

jdeschenes commented:
I should also mention that you found all the issues of string_types. So there shouldn't be any more of these problems.

mdickinson commented:
Apologies for the long silence on this; "real" work got in the way, unfortunately.

I've merged master into this branch (some conflicts, mostly trivial; the only significant conflict was in `get_defs` in `has_traits.py`. I'm intending to get this finished and merged this week.

mdickinson commented on examples/tutorials/tutor.py:
```diff
@@ -766,7 +769,7 @@ def run_code(self):
 
                 # Execute the current lab code:
-                exec module[2:] in values, values
+                exec(module[2:] in values, values)
```
This should be `exec(module[2:], values, values)`. Fixing shortly ...

mdickinson commented on setup.py:
```diff
@@ -171,17 +171,16 @@ def additional_commands():
-        use_2to3=True,
-        use_2to3_fixers=['fixers'],
+        use_2to3=False,
         # traits_listener.ListenerItem has a trait *next* which gets
```
orphaned comment; will remove

mdickinson commented on traits/has_traits.py:
```diff
@@ -1964,7 +1947,7 @@ def _trait_view ( cls, name, view_element, default_name, view_elements,
                 if (result is None) and (handler is not None):
                     method = getattr( handler, name, None )
-                    if callable( method ):
+                    if six.callable( method ):
```
I'll revert these changes; we don't need to support Python 3.0 or 3.1 (indeed, we don't support anything older than Python 3.5 at this point).

mdickinson commented on traits/tests/test_file.py:
```diff
@@ -0,0 +1,51 @@
+import os.path as osp
```
This is a bit nonstandard; changing to just `import os` (with corresponding changes below).

mdickinson commented on traits/tests/test_listeners.py:
```diff
@@ -37,7 +38,7 @@ def captured_stderr():
 
     """
-    new_stderr = cStringIO.StringIO()
+    new_stderr = sm.cStringIO()
```
changing to just `six.StringIO` for consistency with similar uses elsewhere in the codebase.

mdickinson commented on traits/tests/test_new_notifiers.py:
```diff
@@ -26,7 +27,7 @@ def test_notification_on_separate_thread(self):
 
         def on_foo_notifications(obj, name, old, new):
-            thread_id = thread.get_ident()
+            thread_id = sm._thread.get_ident()
```
Taking Python 3's hint that `_thread` should be considered private, and replacing this with `threading.current_thread().ident` (which should work out of the box on both Python 2 and Python 3). Same again below.

mdickinson commented on traits/tests/test_trait_change_event_tracer.py:
```diff
@@ -1,4 +1,7 @@
 """ Tests for the trait change event tracer. """
+
+import six
```
Removing; it looks as though this is unused.

mdickinson commented on traits/traits_listener.py:
```diff
@@ -31,6 +31,8 @@
 from string import whitespace
 from types import MethodType
 
+import six
```
Looks unused. Removing.

mdickinson commented on traits/util/tests/test_async_trait_wait.py:
```diff
@@ -3,6 +3,9 @@
 import time
 import unittest
 
+import six
```
Removing unused import.
